core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp

```diff
@@ -134,27 +134,28 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
     ValueInit::init(functor, &host_result);
     q.memcpy(result_ptr, &host_result, sizeof(host_result));
 
-    q.submit([functor, policy, result_ptr](cl::sycl::handler& cgh) {
-      // FIXME_SYCL a local size larger than 1 doesn't work for all cases
-      cl::sycl::nd_range<1> range(policy.end() - policy.begin(), 1);
-
-      constexpr value_type identity{};
-
-      auto reduction =
-          cl::sycl::ONEAPI::reduction(result_ptr, identity, std::plus<>());
-
-      cgh.parallel_for(
-          range, reduction, [=](cl::sycl::nd_item<1> item, auto& sum) {
-            const typename Policy::index_type id = item.get_global_id(0);
-            value_type partial                   = identity;
-            if constexpr (std::is_same<WorkTag, void>::value)
-              functor(id, partial);
-            else
-              functor(WorkTag(), id, partial);
-            sum.combine(partial);
-          });
-    });
-
+    if (policy.begin() != policy.end()) {
```

Comment:
This still does too much work.  I'm mainly concerned about the allocating, initializing and freeing `result_ptr`, since we don't need it for a zero-length reduction.

Instead, how about something more along the lines of

```
template <typename PolicyType, typename Functor>
  void sycl_direct_launch(const PolicyType& policy, const Functor& functor) const {
    if (policy.begin() == policy.end()) {
        *m_result_ptr = value_type{};
        return;
    }
    // the rest of the body
}
```

Reply:
Yes, good idea. I put it in `execute` now. I still believe that we need to call `ValueInit::init()` so we potentially want to execute that on the host and need to copy it back to the device.